instadeploy.py

```diff
@@ -0,0 +1,234 @@
+        subprocess.call(('rm -R ' + os.path.join(deployment_dir, name, "*")), shell=True)
+
+
+def ensure_mono():
+    if os.path.isdir(ROOT_DIR + 'mono'):
+        return
+    else:
+        subprocess.check_call((
+            'wget',
+            '-q',
+            'http://cdn.mendix.com/mx-buildpack/mono-3.10.0.tar.gz',
+            '-O', ROOT_DIR + 'mono.tar.gz'
+        ))
+        subprocess.check_call((
+            'tar',
+            'xzf',
+            ROOT_DIR + 'mono.tar.gz',
+            '-C', ROOT_DIR
+        ))
+        subprocess.call(('rm', ROOT_DIR + 'mono.tar.gz'))
+
+
+def runmxbuild(project_dir, runtime_version):
+    before = time.time()
+    body = {'args': ['--target=pages', '--loose-version-check', get_mpr_file(project_dir)]}
+    headers = {
+        'Content-Type': 'application/json',
+    }
+
+    body = json.dumps(body)
+    req = urllib2.Request(
+        url='http://localhost:6666/',
+        data=body,
+        headers=headers,
+    )
+    response = urllib2.urlopen(req, timeout=10)
+    if (response.getcode() != 200):
+        raise Exception('http error' + str(response.getcode()))
+    print 'MxBuild compilation succeeded'
+    print 'MxBuild took', time.time() - before, 'seconds'
+
+
+def build(mpk_file, ticker):
+    subprocess.check_call(('unzip', '-oqq', mpk_file, '-d', project_dir))
+    print 'unzip', ticker.next()
+    print 'runtime_version', ticker.next()
+    runmxbuild(project_dir, runtime_version)
+    print 'mxbuild', ticker.next()
+    apply_changes()
+    print 'apply new changes', ticker.next()
+    reload_model()
+    print 'reload model', ticker.next()
+
+
+def reload_model():
+    PASSWORD = os.environ.get('ADMIN_PASSWORD')
+    headers = {
+        'Content-Type': 'application/json',
+        'X-M2EE-Authentication': b64encode(PASSWORD),
+        'Authorization': 'Basic ' + b64encode('MxAdmin:' + PASSWORD),
+    }
+
+    body = {"action": "reload_model"}
+    body = json.dumps(body)
+    req = urllib2.Request(
+        url='http://localhost:' + str(get_admin_port()) + '/',
+        data=body,
+        headers=headers,
+    )
+    response = urllib2.urlopen(req, timeout=5)
+    response_headers = response.info()
+    parsed_response = json.load(response)
+    if (response.getcode() != 200):
+        raise Exception('http error' + str(response.getcode()))
+    print 'RELOADED MODEL GREAT SUCCESS!!'
+
+
+def ticker():
+    prev = time.time()
+    while True:
+        last = time.time()
+        yield last - prev
+        prev = last
+
+def start_mxbuild_server():
+    env = dict(os.environ)
+    env['LD_LIBRARY_PATH'] = os.path.join('lib', 'mono-lib')
+    subprocess.check_call([
+        'sed',
+        '-i',
+        's|/app/vendor/mono/lib/libgdiplus.so|%s|g' % os.path.join(
+            'lib', 'mono-lib', 'libgdiplus.so'
+        ),
+        os.path.join('mono', 'etc', 'mono', 'config'),
```

Comment:
Are we ever going to support windows? (yes this is rhetorical) lib/mono-lib/ligdiplus.so is more readable and there is no user input, so IMHO no need for os.path.join